### PR TITLE
For #2: on the dataset 'winequality.csv'

### PR DESCRIPTION
**Pull request by KaairaGupta**
For #2 
I have attached the results with one of the models.

results

### Objectives:

- To compare several methods of classification on the dataset `winequality.csv`
- To take care of the pre-processing of data, as seen by the EDA that this data needs some tweaks.

The key takeaways from the data are mentioned inline in [this](https://github.com/KaairaGupta/PRESC/blob/master/dev/KaairaGupta/data_visualisation.ipynb) file, which have been taken care of [here](https://github.com/KaairaGupta/PRESC/blob/master/dev/KaairaGupta/data_loader.py)

I have undersampled the data in this PR [here](https://github.com/KaairaGupta/PRESC/blob/master/dev/KaairaGupta/data_loader.py) and am currently working on more methods for taking care of the _imbalanced data_ that this data is.
I have changed the evaluation metrices for the same reason, using False positives, True positives etc.
Since the quality column is directly proportional to the recommend column, I have removed the quality column. Further, I want to work on predicting `quality` using recommend and other parameters but since this was a classification problem, I wanted to know if I can take that up? @dzeber Please let me know if I can.

 ### Results:
- I have written a [model class](https://github.com/KaairaGupta/PRESC/blob/master/dev/KaairaGupta/Ml_functions.py) which takes care of confusion matrix, Classification records etc to take care of the comparative study that we have to carry out
- Used [several classifiers with several different normal transactions]() to compare the metrices.

I think it would be appropriate for me to make an md file of all the comparative results that I got from this study and the methods, if you approve of my contribution. Please suggest the improvements so that I can work on the md file with detailed report thereafter.  

Further:
- I have made a folder as suggested.
- I have also added `.vscode/` in the gitignore for the cache of the IDE

### CODE REVIEW

File: dev/KaairaGupta/Ml_functions.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+
+from data_loader import load_split_preprocessed_data
+from sklearn.metrics import confusion_matrix,recall_score,precision_recall_curve,auc,roc_curve,roc_auc_score,classification_report
+import warnings
+import matplotlib.pyplot as plt # to plot graph
+import seaborn as sns
+warnings.filterwarnings('ignore')
+
+#model function
+def model(model,features_train,features_test,labels_train,labels_test):
```
Comment (mlopatka) [inline dev/KaairaGupta/Ml_functions.py]: nit: Please include a full doc-string explaining the contents of `model`

File: dev/KaairaGupta/data_loader.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import numpy as np
+from sklearn.model_selection import train_test_split
+from sklearn.preprocessing import StandardScaler
+from sklearn.discriminant_analysis import LinearDiscriminantAnalysis as LDA
+
+
+
+def load_split_preprocessed_data(i):
+
+    data = pd.read_csv("../../datasets/winequality.csv")
```
Comment (mlopatka) [inline dev/KaairaGupta/data_loader.py]: Please specify the local filepath as a variable outside this call for easier local testabilitiy.

File: dev/KaairaGupta/data_loader.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    Count_Fraud_transacation = len(data["recommend"]==True)
```
Comment (mlopatka) [inline dev/KaairaGupta/data_loader.py]: Please adhere to python naming conventions as per PEP8 (this may also be enforced by black)

File: dev/KaairaGupta/data_loader.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def load_split_preprocessed_data(i):
+
+    data = pd.read_csv("../../datasets/winequality.csv")
+    Count_Fraud_transacation = len(data["recommend"]==True) 
+    fraud_indices= np.array((data.recommend==True).index)
+    normal_indices = np.array((data.recommend==False).index)
+
+    undersample_data = undersample(normal_indices,fraud_indices,i,Count_Fraud_transacation,data)
```
Comment (mlopatka) [inline dev/KaairaGupta/data_loader.py]: What is the meaning of the variable name `Count_Fraud_transacation`?
In the context of the wine dataset this variable naming is confusing.

File: dev/KaairaGupta/data_loader.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+    # since data is skewed, we add epsilon to data and take its logarithm
+    undersample_data[['fixed acidity','volatile acidity','residual sugar','chlorides','free sulfur dioxide','total sulfur dioxide','density','sulphates','alcohol','quality']] = np.add(undersample_data[['fixed acidity','volatile acidity','residual sugar','chlorides','free sulfur dioxide','total sulfur dioxide','density','sulphates','alcohol','quality']],0.00000001)
```
Comment (mlopatka) [inline dev/KaairaGupta/data_loader.py]: What is the basis of the parameter selection for `0.00000001`?
Please break this out as a variable and document the motivation for hard coding a specific value.

File: dev/KaairaGupta/data_loader.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    undersample_data[['fixed acidity','volatile acidity','residual sugar','chlorides','free sulfur dioxide','total sulfur dioxide','density','pH','sulphates','alcohol','quality']] = np.log(undersample_data[['fixed acidity','volatile acidity','residual sugar','chlorides','free sulfur dioxide','total sulfur dioxide','density','pH','sulphates','alcohol','quality']])
+    undersample_data.drop(["quality"],axis=1,inplace=True)
+    x_features= undersample_data.drop(["recommend"],axis=1,inplace=False)
+    x_labels=undersample_data["recommend"]
+
+    x_features_train,x_features_test,x_labels_train,x_labels_test = train_test_split(x_features,x_labels,test_size=0.3)
```
Comment (mlopatka) [inline dev/KaairaGupta/data_loader.py]: what is the motivation for a test split at exactly `test_size=0.3`?
Does the performance observed vary with the train/test split ratio? 
Please document the motivation for hard coding a specific value.

File: dev/KaairaGupta/main.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
+from data_loader import load_split_preprocessed_data
+from Ml_functions import model
+from sklearn.ensemble import RandomForestClassifier # Random forest classifier
+from sklearn.tree import DecisionTreeClassifier # for Decision Tree classifier
+from sklearn.svm import SVC # for SVM classification
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import train_test_split # to split the data
+from sklearn.model_selection import KFold # For cross vbalidation
```
Comment (mlopatka) [inline dev/KaairaGupta/main.py]: nit: typo in `vbalidation`

File: dev/KaairaGupta/main.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+from sklearn.model_selection import GridSearchCV # for tunnig hyper parameter it will use all combination of given parameters
+from sklearn.model_selection import RandomizedSearchCV # same for tunning hyper parameter but will use random combinations of parameters
+
+#take normal transaction in  0.5 , 0.66 and 0.75  proportion of total data
```
Comment (mlopatka) [inline dev/KaairaGupta/main.py]: please extend the verbosity of this comment, I am not sure what a `transaction` refers to hear.

File: dev/KaairaGupta/Ml_functions.py (hunk 1 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/Ml_functions.py]: I have included as required. Also, commented most of the code.
Thanks!

File: dev/KaairaGupta/data_loader.py (hunk 2 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/data_loader.py]: Done!

File: dev/KaairaGupta/data_loader.py (hunk 3 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/data_loader.py]: Yes, sorry for this. Made changes in the entire files.

File: dev/KaairaGupta/data_loader.py (hunk 4 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/data_loader.py]: Renamed the variables for better intution

File: dev/KaairaGupta/data_loader.py (hunk 5 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/data_loader.py]: Commented the required statements. Please check. Thanks!

File: dev/KaairaGupta/data_loader.py (hunk 6 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/data_loader.py]: Done this as well. PTAL

File: dev/KaairaGupta/main.py (hunk 7 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/main.py]: Thanks! Changed this.

File: dev/KaairaGupta/main.py (hunk 8 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/main.py]: Yes. I have made the comment more intuitive now

**Comment by KaairaGupta** [comment]
@mlopatka Thanks for the reviews! I have made the necessary changes. Please review.

**Comment by KaairaGupta** [comment]
@mlopatka PTAL 
Thanks!

**Comment by KaairaGupta** [comment]
@mlopatka I have added classification, accuracy, etc in ipynb notebook for easier visualisation as per your reviews. Please have a look. Thanks!

File: dev/KaairaGupta/winequality/data_loader.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+
+
+def load_split_preprocessed_data(i):
+
+    data_path = "../../../datasets/winequality.csv"
+    data = pd.read_csv(data_path)
+    count_false_recommendation = len(data["recommend"]==True) 
+    false_indices= np.array((data.recommend==True).index)
+    true_indices = np.array((data.recommend==False).index)
```
Comment (dzeber) [inline dev/KaairaGupta/winequality/data_loader.py]: It looks like the variable names `true_` and `false_` on lines 13-15 are the opposite of what the code does. Unless I'm missing something, I'd recommend renaming these to something more intuitive.

File: dev/KaairaGupta/winequality/data_visualisation.ipynb [hunk 10]
```diff
@@ -0,0 +1,1421 @@
...
...JCd7boA71m3OSGEOA80s67bKKVMs7NNN+D/rNscFkJcubXTCcv0on9YP1sH7AIygFxghRDiV+AXewlYx9suA1gXNLxcX6/y4tNwCy3qBXEN1ZOXcKl0mbp+pF8pE1JURjhpaffRU8R/t4PENfbvVMf97w8iv3iaqPl2HwWsUHlxqcV6YHQhfqXziUvFxSZnXYievISi6vLt1Y6sI+fIT6m64R72GONT0YXY5BbsR36Cvd2sesqLT0VX17Zu9IXDIG3LuIT6k2l97WLtffPp2prcmCQKUjMASFuzG+/2N5Hy3XYMUbEcH/4KAK6NQ/DtE+mQfG4ExvhUdMWOlZq1z5VkSkjBObiox8op2B9TUmqpcrpmYQS+PIW4R5/FfDmz1PqqYk5KRhsYWPhaG+iPKaX08CinJo3xnTWd1KeewZxhOWZMScmYkpPJP34CAMNv2/EaWTUNuJtH9+GmB3sCkHLoLB42+5h7iJ6cxPRi5XPTMtF5uyO0GqTJjEeInhzreT0/y8Dv04oeOblv19tklnjep7LkxqfhahO7m53rZ258Gq51ba6xIaXL2CO0GoLv6MAffWdXXMDXoc+oAfQc3heAs4ej8LO5nuqD/UhPKjt2aTaz++cd3PHoULZ/s4VL8akcTsskz5BHniGPE38dp8HNjUg4F1/peQB0GNWXyOGW/Sz28Fm8berKO1hPZlK63e26P3k3HnovvppVjW7eYHlurZ01nzg7+WSVkY89bYZ1Y+eSnwG4dD6R9AvJ+DcJIe5Q9ZmwqcJVzSyUSCnXAGtKLLM7m5mUckxFfW6t74HDcl/GXq3mlShjb7uyXG1Qu73PElgafW2tPy2klOOtLfeOWKYWHQqsu8r7/iOXD5zBvXEwbg0CEM5agod2Jmn9vmJlktbvI3SYpbejTmQ4+Zk55FlPFC3ffpSs07FEf1B8Rib3sODC3wP7R5J92jHPj2UejMK1cQguDQIRzk4EDL2dtA3FG5apG/YSeF8PALwimmLKzCHf5sQXeFeXajd8EiD70Glcw0LQ1bfkph/ShfSNVfNw+z+RdTAKt7AQXKzx+w/pQtr6vcXKXFq/h4Bh3QHwjGhKgbVu8mJT8IpshsZNB0CdLq3IOW2Zac7Zz/q8ixDUm3IviSs3OC6pWi7rYBSuNnWmH9KFSxuqz5DC8so9ehLnhnVxqhsEzk54DuxB9m+7i
```
Comment (dzeber) [inline dev/KaairaGupta/winequality/data_visualisation.ipynb]: Did you mean `0.5, 0.33 and 0.25 proportions` for `true`? That's what the code does as far as I can tell. Eg. when `i = 2` it looks like you are retaining twice as many `false` cases as `true`. It would be great to use these same proportion numbers in the output below rather than `1 proportion`, `2 proportion`, etc.

File: dev/KaairaGupta/winequality/data_loader.py (hunk 9 above)
Comment (KaairaGupta) [reply dev/KaairaGupta/winequality/data_loader.py]: Yes, it was a silly mistake from my side..I have taken care of it and am submitting the changes.

**Comment by KaairaGupta** [comment]
@mlopatka I have made all the requested changes. Renamed the variables, ran python black, and added doc strings. Please review. Thanks

**Comment by dzeber** [comment]
Looks great! Thanks for following up on all of these comments!
